src/server/socket.js

```diff
@@ -1,103 +1,204 @@
+    whenSimHostReady.resolve();
+
+    setupSimHostHandlers();
+
+    whenAppHostConnected.promise
+        .then(onAppHostConnected);
+}
+
+function handleAppPluginList(data) {
+    var socket = hostSockets.SIM_HOST;
 
-            // Set up telemetry if necessary.
-            if (config.telemetry) {
-                socket.emit('init-telemetry');
+    socket.once('start', handleStart);
+
+    emitToHost(SIM_HOST, 'app-plugin-list', data);
+}
+
+function handleStart() {
+    emitToHost(APP_HOST, 'start');
+}
+
+function setupSimHostHandlers() {
+    log.log('Setup handlers for SIM_HOST');
+
+    var socket = hostSockets.SIM_HOST;
+
+    socket.on('exec-success', function (data) {
+        emitToHost(APP_HOST, 'exec-success', data);
+    });
+    socket.on('exec-failure', function (data) {
+        emitToHost(APP_HOST, 'exec-failure', data);
+    });
+
+    socket.on('plugin-message', function (data) {
+        emitToHost(APP_HOST, 'plugin-message', data);
+    });
+
+    socket.on('plugin-method', function (data, callback) {
+        emitToHost(APP_HOST, 'plugin-method', data, callback);
+    });
+
+    socket.on('telemetry', function (data) {
+        telemetry.handleClientTelemetry(data);
+    });
+
+    // Set up telemetry if necessary.
+    if (config.telemetry) {
+        socket.emit('init-telemetry');
+    }
+
+    handlePendingEmits(SIM_HOST);
+
+}
+
+function init(server) {
+    var io = require('socket.io')(server);
+
+    io.on('connection', function (socket) {
+        socket.on('register-app-host', function () {
+            log.log('APP_HOST connected to the server');
+            if (hostSockets[APP_HOST]) {
+                log.log('Overriding previously connected APP_HOST');
+                resetAppHostState();
+            }
+            hostSockets[APP_HOST] = socket;
+            whenSimHostReady.promise
+                .then(onSimHostReady);
+            whenAppHostConnected.resolve();
+        });
+
+        socket.on('register-simulation-host', function () {
+            log.log('SIM_HOST connected to the server');
+            if (hostSockets[SIM_HOST]) {
+                log.log('Overriding previously connected SIM_HOST');
+                resetSimHostState();
+            }
+            hostSockets[SIM_HOST] = socket;
+            handleSimHostRegistration(socket);
+        });
+
+        socket.on('disconnect', function () {
+            var type;
+            Object.keys(hostSockets).forEach(function (t) {
+                if (hostSockets[t] === socket) {
+                    type = t;
+                }
+            });
+            if (!type) {
+                log.log('Disconnect for an inactive socket');
+                return;
+            }
+            hostSockets[type] = undefined;
+            log.log(type + ' disconnected to the server');
+            switch (type) {
+                case APP_HOST:
+                    resetAppHostState();
```

Comment:
When app-host is disconnected we reset the state, that means creating a new promise and attach a callback for when the promise is resolved, but when the app-host sends the `register-app-host` message, the promise is resolved. And the handler gets executed assuming that there's a connected sim-host instance.